### PR TITLE
monocle3, scala3 upgrade

### PR DESCRIPTION
**Pull request by inanme**
I guess it's customary to create a PR for Scala3 & Monocle3 upgrades. This PR removes all deprecated Monocle API usages.

### CODE REVIEW

File: optics/src/main/scala/io/circe/optics/JsonNumberOptics.scala
Comment (inanme) [inline optics/src/main/scala/io/circe/optics/JsonNumberOptics.scala:42]: stop scala3 compiler warning, also applies to other occurrences.

**Comment by inanme** [comment]
@zmccoy @zarthross please let me what you think, thanks

**Comment by zmccoy** [comment]
Hi @inanme at first glance we'll want keep the Scala versions at the Build level such as : https://github.com/circe/circe/blob/series/0.14.x/build.sbt#L13.  That will keep 2.12,2.13, and 3 in github action workflow yaml.  

You'll probably want to rebase as well as there were some merges from a few weeks ago that added in the sbt-circe-org plugin which helps make things a bit cleaner.

**Comment by inanme** [comment]
>>> You'll probably want to rebase
rebase done.

File: optics/src/main/scala/io/circe/optics/JsonPath.scala
Comment (GreyPlane) [inline optics/src/main/scala/io/circe/optics/JsonPath.scala:123]: same reason as above, written as `FilterIndex.filterIndex[Vector[Json], Int, Json](p)`

**Review by GreyPlane — APPROVED** [review]
just few minor style issues, looks good!

File: optics/src/main/scala/io/circe/optics/JsonPath.scala
Comment (inanme) [reply optics/src/main/scala/io/circe/optics/JsonPath.scala:123]: thanks, agreed, https://github.com/circe/circe-optics/pull/336/commits/3e2abe342676de522e149169bdb7120c6a1ae87a

File: optics/src/main/scala/io/circe/optics/JsonPath.scala
Comment (inanme) [reply optics/src/main/scala/io/circe/optics/JsonPath.scala:117]: thanks, agreed, https://github.com/circe/circe-optics/pull/336/commits/3e2abe342676de522e149169bdb7120c6a1ae87a

File: optics/src/main/scala/io/circe/optics/JsonPath.scala
Comment (inanme) [reply optics/src/main/scala/io/circe/optics/JsonPath.scala]: also  https://github.com/circe/circe-optics/pull/336/commits/844a837da9be1aa2f0106e19caf416ee50f281ba

File: optics/src/main/scala/io/circe/optics/JsonPath.scala
Comment (inanme) [reply optics/src/main/scala/io/circe/optics/JsonPath.scala]: also https://github.com/circe/circe-optics/pull/336/commits/844a837da9be1aa2f0106e19caf416ee50f281ba

**Comment by zmccoy** [comment]
@inanme The dropping of 2.12 for monocle3 is not an issue in that case, I didn't know they dropped 2.12.   We're going to want to think about branching strategy and version numbers expectations as this is bin incompat change.

**Comment by zmccoy** [comment]
The failures right now are expected as mima is showing that things are not bin compat.  We can solve this once we figure out our branch / version strategy for this then :)

**Comment by GreyPlane** [comment]
I didn't found any particular reason for Monocle to drop the support for Scala 2.12.x, Only because they're trying to reduce maintenance burden so decided to only support 2 major versions of Scala, I think we could ask them if it is possible to bring back it.

**Comment by inanme** [comment]
Would this configuration work in circe ecosystem? I setup projects like this using `sbt-projectmatrix`.
0.14.2 -> (scala212 -> monocle2) -> `circe-optics` / /BAU
0.14.2 -> (scala213 -> monocle2) -> `circe-optics` / /BAU
0.14.2 -> (scala213 -> monocle3) -> `circe-optics-monocle3` // new artefact
0.14.2 -> (scala3 -> monocle3) -> `circe-optics` // new artefact

**Comment by zarthross** [comment]
@zmccoy and I discussed, and we think the best approach would be to use this naming scheme:
```
0.14.2 -> (scala212 -> monocle2) -> circe-optics
0.14.2 -> (scala213 -> monocle2) -> circe-optics
0.14.2 -> (scala213 -> monocle3) -> circe-optics-monocle_3 
0.14.2 -> (scala3   -> monocle3) -> circe-optics-monocle_3
```

We'll want to add something to the readme as well, to point users to this new artifact.

@GreyPlane How does this approach sound to you?

**Comment by GreyPlane** [comment]
Personally, I'd like to say continuously support for circe-optics(monocle2) seems to unnecessary since Monocle itself dropped the support already, so release to next major version looks fine to me, and even Circe itself's major release holds bin comp, but it's more like a bonus since semantic version doesn't enforce that, so even at some point one of the Circe's associated project doesn't keep it, this behavior should be expected with some extent.

But I don't hold very strong opinion about this, especially if there is real world needs for using circe-optics with monocle 2, I'm OK with the new artifact.

**Comment by GreyPlane** [comment]
@inanme I guess you have to exclude mima previous artifact for Scala 3 cross build as [document](https://github.com/lightbend/mima#setting-different-mimapreviousartifacts) says.

@zarthross and so, are we decided to release new artifact under a new name?

**Comment by QuitHub** [comment]
It would be really great to have this available!

**Comment by zarthross** [comment]
@GreyPlane, I think you have a persuasive argument.  I'm ok with abandoning support for monocle 2, releasing a new version of circe-optics as 0.15 with monocle 3, and including in the readme a table with a note about the versioning and how it relates to circe-core.  See https://github.com/circe/circe-generic-extras#versioning please use the same verbiage and structure for consistency.

**Comment by joan38** [comment]
Hey, Thanks for the work here!
We are also blocked on this to upgrade to Scala 3. Sounds like depending on the 2.13 package from Scala 3 is not working.

**Comment by GreyPlane** [comment]
@inanme Hi buddy, I am currently un-employed, so I can pick this up as I have bit more time to focus. If you like me to look more into this issue.

**Comment by joan38** [comment]
```
ResolveException: Error downloading io.circe:circe-optics_3:0.14.1
```
Why is it trying to resolve itself?

**Comment by joan38** [comment]
So I understand the current build is not going through because of MiMa.
Is there anything we can do to move forward? Sounds like we have a sizable amount of people volunteering. Is the issue technical or political here?
We have projects we need to move forward on Scala 3 depending on circe-optics. Can a maintainer step forward?

**Comment by GreyPlane** [comment]
@joan38 not any problem really, I think we have reach the consensus here and there isn't any hard technical issue, just have to finish the work.
I guess it's ok to pick this work into new branch and submit a new PR with respect to the original, you'd like to do it by yourself? or i can do it as i said before.

**Comment by zarthross** [comment]
Thanks for the work on this.

**Pull request closed without merging**
